### PR TITLE
Update quest lines, Refactor quests

### PR DESCRIPTION
**Pull request by RedSparr0w**
Refactored quests to `App.game.quests`, no longer saved in `player`.
Moved the different quest types into their own folder.
QuestLines has been reworked a little bit.
image
Added a small (i) icon that shows the quest line description when hovered.
image
Added ability for custom quest rewards (only for custom quests).
Added ability for gym reward functions (currently only used to start Deoxys quest line)
Deoxys no longer roaming, awarded at the end of the quest.

All previous save data should convert over automatically
  - Current active quests
  - Current quest progress
  - Completed quest
  - Quest xp/level

### CODE REVIEW

File: src/scripts/quests/Quests.ts [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    public questLines: KnockoutObservableArray<QuestLine> = ko.observableArray();
+
+    constructor() {}
+
+    getQuestLine(name) {
+        return this.questLines().find(ql => ql.name.toLowerCase() == name.toLowerCase());
+    }
+
+    get level(): KnockoutComputed<number> {
+        return ko.pureComputed((): number => {
```
Comment (Aegyo) [inline src/scripts/quests/Quests.ts]: I think all of these `get thing(): KnockoutComputed<T>` are going to be creating a new computed every time they are called. Is there some reason we don't want to just 
```
public level: KnockoutComputed<number> = ko.pureComputed(...
```
which would only make the computed once?

File: src/scripts/quests/Quests.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        return App.game.wallet.hasAmount(this.getRefreshCost());
+    }
+
+    // Returns 0 when all quests are complete, ~1 million when none are
+    public getRefreshCost(): Amount {
+        const notComplete = this.incompleteQuests().length;
+        return new Amount(Math.floor(250000 * Math.LOG10E * Math.log(Math.pow(notComplete, 4) + 1)), GameConstants.Currency.money);
+    }
+
+    public getCompletedQuest
```
Comment (Aegyo) [inline src/scripts/quests/Quests.ts]: What is this for?

**Review by Aegyo — CHANGES_REQUESTED** [review]
The tutorial questline is restarting from the beginning if you refresh the page. In develop, it is only making you restart the current quest within the tutorial questline.
If anyone gets past the key item buying quest, but leaves before finishing the tutorial, they will be stuck and have to restart (because you can't buy the key item a second time)

File: src/scripts/quests/Quests.ts (hunk 1 above)
Comment (RedSparr0w) [reply src/scripts/quests/Quests.ts]: Yeah, that makes a lot more sense

File: src/scripts/quests/Quests.ts (hunk 2 above)
Comment (RedSparr0w) [reply src/scripts/quests/Quests.ts]: 🤔 _Nothing, I think.._ 
Gone 💨

**Comment by RedSparr0w** [comment]
I wasn't able to reproduce the tutorial not re-loading from the current point:
reload

Was there a certain point you reached where it was happening?

**Comment by Aegyo** [comment]
I went as far as moving to route 2, defeated a couple of pokemon, waited for the console to say it had saved, and then refreshed
When it had reloaded, it had me back on the first quest
quest

**Comment by RedSparr0w** [comment]
Hmm, are you able to add a console.log or something to see what's being passed through the save when it happens?

I've updated the dungeon ticket quest to always start from `0` so if you already have it, it will auto complete, so you won't be soft locked.
But it's still not ideal that it's happening at all 🤔

**Review by Aegyo — APPROVED** [review]
Note to self: restart gulp if there is a new version update...

**Pull request merged**